app/lib/shared/utils.dart

```diff
@@ -361,3 +362,38 @@ bool fixedTimeEquals(String a, String b) {
   }
   return result == 0;
 }
+
+extension StreamBoundedForEach<T> on Stream<T> {
+  /// Executes items using [eachFn] with maximum [concurrency].
+  Future<void> boundedForEach(
+    int concurrency,
+    FutureOr<void> Function(T) eachFn,
+  ) async {
+    final pool = Pool(concurrency);
+    final futures = Queue<Future>();
+    // keeps the size of the futures limited, allows some difference in the run times
+    final maxQueueSize = concurrency * 2;
+    try {
+      await for (final item in this) {
+        final f = pool.withResource(() => eachFn(item));
+        futures.add(f);
+        while (futures.length > maxQueueSize) {
+          await futures.removeFirst();
```

Comment:
Shouldn't this be a `Future.race` ?

Reply:
Not sure: what's that, in which library?

Reply:
I'm not sure if that's what we want:

> The returned future is completed with the result of the first future in futures to report that it is complete, whether it's with a value or an error. The results of all the other futures are discarded.